### PR TITLE
feature(whl): add gpt utils

### PR DESCRIPTION
**Pull request by kxzxvbk**
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/torch_utils/optimizer_helper.py [hunk 1]
```diff
@@ -730,3 +731,59 @@ def _retrieve_grad(self):
...
+            elif pn.endswith('weight') and isinstance(m, whitelist_weight_modules):
+                # weights of whitelist modules will be weight decayed
+                decay.add(fpn)
+            elif pn.endswith('weight') and isinstance(m, blacklist_weight_modules):
+                # weights of blacklist modules will NOT be weight decayed
+                no_decay.add(fpn)
+            else:
+                decay.add(fpn)
+
+    decay = decay - no_decay
```
Comment (karroyan) [inline ding/torch_utils/optimizer_helper.py:767]: With this line, the validation of len(inter_params) == 0 in L772 will not be effective?

File: ding/torch_utils/tests/test_backend_helper.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    def test_tf32(self):
+        r"""
+        Overview:
+            Test the tf32.
+        """
+        enable_tf32()
+        net = torch.nn.Linear(3, 4)
+        x = torch.randn(1, 3)
+        y = torch.sum(net(x))
+        y.backward()
```
Comment (karroyan) [inline ding/torch_utils/tests/test_backend_helper.py:19]: Add some assert in this test

File: ding/torch_utils/tests/test_optimizer.py [hunk 3]
```diff
@@ -104,21 +104,21 @@ def try_optim_with(tname, t, optim_t):
...
-            'grad_clip': [None, 'clip_momentum', 'clip_value', 'clip_norm', 'clip_momentum_norm'],
-            'grad_norm': [None],
-            'grad_ignore': [None, 'ignore_momentum', 'ignore_value', 'ignore_norm', 'ignore_momentum_norm'],
-        }
-
-        for optim_t in support_type['optim']:
-            for tname in ['grad_clip', 'grad_ignore']:
-                for t in support_type[tname]:
-                    try_optim_with(tname=tname, t=t, optim_t=optim_t)
+# @pytest.mark.unittest
```
Comment (karroyan) [inline ding/torch_utils/tests/test_optimizer.py:107]: Why comment out this test?

File: ding/torch_utils/tests/test_backend_helper.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import pytest
+import torch
+
+from ding.torch_utils.backend_helper import enable_tf32
+
+
+@pytest.mark.unittest
```
Comment (PaParaZz1) [inline ding/torch_utils/tests/test_backend_helper.py:7]: cudatest

File: ding/torch_utils/tests/test_model_helper.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+@pytest.mark.unittest
+class TestModelHelper:
+
+    def test_model_helper(self):
+        r"""
+        Overview:
+            Test the model helper.
+        """
+        net = torch.nn.Linear(3, 4, bias=False)
+        assert get_num_params(net) == 12
```
Comment (PaParaZz1) [inline ding/torch_utils/tests/test_model_helper.py:16]: also test a conv case

File: ding/torch_utils/network/activation.py [hunk 6]
```diff
@@ -72,6 +74,15 @@ def forward(self, x):
         return x
 
 
+class GELU(nn.Module):
+
+    def __init__(self):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/activation.py:79]: add paper link and overview comments, typing lint

File: ding/torch_utils/lr_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
+from functools import partial
+import math
+
+from torch.optim.lr_scheduler import LambdaLR
+
+
+def get_lr(it, warmup_epochs, learning_rate, lr_decay_epochs, min_lr):
```
Comment (PaParaZz1) [inline ding/torch_utils/lr_scheduler.py:7]: typing lint

File: ding/torch_utils/optimizer_helper.py [hunk 8]
```diff
@@ -730,3 +731,58 @@ def _retrieve_grad(self):
                 grad.append(p.grad.clone())
                 has_grad.append(torch.ones_like(p).to(p.device))
         return grad, shape, has_grad
+
+
+def configure_weight_decay(model, weight_decay):
+    """
```
Comment (PaParaZz1) [inline ding/torch_utils/optimizer_helper.py:737]: modify comments like our format

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**